apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -0,0 +1,606 @@
+  end
+
+  def state do
+    GenServer.call(__MODULE__, :state)
+  end
+
+  @doc """
+  Starts a synchronizing process between our node and the node of the given peer_pid
+  """
+  @spec start_sync(pid(), binary()) :: :ok | {:error, String.t()}
+  def start_sync(peer_pid, remote_hash) do
+    GenServer.call(__MODULE__, {:start_sync, peer_pid, remote_hash})
+  end
+
+  @doc """
+  Fetches the remote Pool of transactions
+  """
+  @spec fetch_mempool(pid()) :: :ok | {:error, String.t()}
+  def fetch_mempool(peer_pid) do
+    GenServer.call(__MODULE__, {:fetch_mempool, peer_pid})
+  end
+
+  @doc """
+  Checks weather the sync is in progress
+  """
+  @spec sync_in_progress?(pid()) :: false | {true, non_neg_integer}
+  def sync_in_progress?(peer_pid) do
+    GenServer.call(__MODULE__, {:sync_in_progress, peer_pid})
+  end
+
+  @spec new_peer?(pid(), Header.t(), non_neg_integer(), binary()) :: true | false
+  def new_peer?(peer_pid, header, agreed_height, hash) do
+    GenServer.call(__MODULE__, {:is_new_peer, self(), peer_pid, header, agreed_height, hash})
+  end
+
+  @doc """
+  Does the fetching of the blocks depending on the hash_pool.
+  """
+  @spec fetch_next(pid(), non_neg_integer(), binary(), any()) :: :done | tuple()
+  def fetch_next(peer_pid, height_in, hash_in, result) do
+    GenServer.call(__MODULE__, {:fetch_next, peer_pid, height_in, hash_in, result}, 30_000)
+  end
+
+  @spec update_hash_pool(list()) :: list(Header.t())
+  def update_hash_pool(hashes) do
+    GenServer.call(__MODULE__, {:update_hash_pool, hashes})
+  end
+
+  def delete_from_pool(peer_pid) do
+    GenServer.cast(__MODULE__, {:delete_from_pool, peer_pid})
+  end
+
+  ## INNER FUNCTIONS ##
+
+  def handle_call(:state, _from, state) do
+    {:reply, state, state}
+  end
+
+  def handle_call({:start_sync, peer_pid, remote_hash}, _from, state) do
+    :ok = Jobs.enqueue(:sync_jobs, {:start_sync, peer_pid, remote_hash})
+    {:reply, :ok, state}
+  end
+
+  def handle_call({:fetch_mempool, peer_pid}, _from, state) do
+    :ok = Jobs.enqueue(:sync_jobs, {:fetch_mempool, peer_pid})
+    {:reply, :ok, state}
+  end
+
+  def handle_call(
+        {:is_new_peer, process_pid, peer_pid, header, agreed_height, hash},
+        _from,
+        %{sync_pool: pool} = state
+      ) do
+    height = header.height
+    difficulty = Scientific.target_to_difficulty(header.target)
+
+    {new?, new_pool} =
+      try_add_peer(
+        Sync.new(%{
+          difficulty: difficulty,
+          from: agreed_height,
+          to: height,
+          hash: hash,
+          peer: peer_pid,
+          pid: process_pid
+        }),
+        pool
+      )
+
+    if new? do
+      Process.monitor(process_pid)
+    else
+      :ok
+    end
+
+    {:reply, new?, %{state | sync_pool: new_pool}}
+  end
+
+  def handle_call({:sync_in_progress, peer_pid}, _from, %{sync_pool: pool} = state) do
+    result =
+      case Enum.find(pool, false, fn peer -> Map.get(peer, :pid) == peer_pid end) do
```

Comment:
Is there some important reason for `pool` to be a list instead of map (pid -> peer)?

Reply:
Maybe not, it was made this way, because that's how `epoch` did it. But I don't find a reason why to change it for the sake of it right now. I feel that if we decide to change it, to do so in an upcoming issue.

Reply:
The goal is to copy specification, not translate erlang code to elixir. The question from @cytadela8 seems appropriate.